java/ql/lib/semmle/code/java/Expr.qll

```diff
@@ -1540,27 +1562,43 @@ class SwitchExpr extends Expr, StmtParent, @switchexpr {
 /** An `instanceof` expression. */
 class InstanceOfExpr extends Expr, @instanceofexpr {
   /** Gets the expression on the left-hand side of the `instanceof` operator. */
-  Expr getExpr() {
-    if this.isPattern()
-    then result = this.getLocalVariableDeclExpr().getInit()
-    else result.isNthChildOf(this, 0)
-  }
+  Expr getExpr() { result.isNthChildOf(this, 0) }
+
+  /**
+   * Gets the pattern of an `x instanceof T pattern` expression, if any.
+   */
+  PatternExpr getPattern() { result.isNthChildOf(this, 2) }
 
   /**
    * Holds if this `instanceof` expression uses pattern matching.
    */
-  predicate isPattern() { exists(this.getLocalVariableDeclExpr()) }
+  predicate isPattern() { exists(this.getPattern()) }
 
   /**
-   * Gets the local variable declaration of this `instanceof` expression if pattern matching is used.
+   * Gets the local variable declaration of this `instanceof` expression if simple pattern matching is used.
+   *
+   * Note that this won't get anything when record pattern matching is used-- for more general patterns,
+   * use `getPattern`.
    */
-  LocalVariableDeclExpr getLocalVariableDeclExpr() { result.isNthChildOf(this, 0) }
+  LocalVariableDeclExpr getLocalVariableDeclExpr() { result = this.getPattern().asBindingPattern() }
 
-  /** Gets the access to the type on the right-hand side of the `instanceof` operator. */
+  /**
+   * Gets the access to the type on the right-hand side of the `instanceof` operator.
+   *
+   * This does not match record patterns, which have a record pattern (use `getPattern`) not a type access.
+   */
   Expr getTypeName() { result.isNthChildOf(this, 1) }
 
-  /** Gets the type this `instanceof` expression checks for. */
-  RefType getCheckedType() { result = this.getTypeName().getType() }
+  /**
+   * Gets the type this `instanceof` expression checks for.
+   *
+   * For a match against a record pattern, this is the type of the outermost record type.
+   */
+  RefType getCheckedType() {
```

Comment:
I think the way that this predicate is typically used means that we can't allow it to hold for restricted record patterns, i.e. those for which `isUnrestricted()` doesn't hold.

Reply:
Done, though most users seemed interested in the positive case (instanceof passes ==> e is of type) not the negative, so most have been updated to notice even restricted record type checks.

Reply:
Hmm, right, I see. In any case, I think this is the safer option.